### PR TITLE
New workflow to get latest PRs in a deployment

### PR DESCRIPTION
**Pull request by sunnyyang99**
## Ticket
https://app.asana.com/0/1201760991851092/1204108171891109/f
## Goal
Agree on the best format for the release message.

## Code Changes
- Get the runs for a given workflow (default to `deploy_prod.yml`, which is the same for all 3 repos)
- Get the timestamp of the commit for the previous run
- Get all commits that occurred strictly after that commit
- Get the PR for each commit
- Assemble a slack message with the title, link, and author of each PR
## How I Know This Works
Screenshot 2023-07-03 at 2 02 46 PM

When I ran this for a workflow with no new releases, there is a different message.
Screenshot 2023-07-03 at 12 36 04 PM

Not sure what will happen if there is a commit w/o a PR, but I don't think we should include individual commits.
## Notes & Questions for Reviewers
Do y'all prefer using code block for the PR title, like how the deploy message is?
code block style

Do we want to modify the deploy prod message now? Possibly redundant to state the sha and PR that was deployed.

I'm also considering making this workflow output the message instead of post to slack, then the `deploy_prod.yml` file would add that to the slack message, so that there is only one deploy message per repo.

Next steps would be integrating this with the API, FE, and ETL project.
### Checks
- [x] Changes in this PR follows Haus' [code contribution guidelines](https://docs.google.com/document/d/1-3pOkN0Qk5OrsoDJCFPL_Jwunc6ptXLzSpTdTRAH7to/edit#heading=h.fy8j7l7xqyx8) or if it breaks guidelines I've explained why it's necessary.
- [ ] I'm making a #major or #minor [semver](https://semver.org/) change and have that [present](https://github.com/jasonamyers/github-bumpversion-action) in a commit message

### CODE REVIEW

File: .github/workflows/get_prs_in_release.yml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        echo "• \`$PR_TITLE\` (<$PR_URL|$PR_NUMBER>) by $PR_AUTHOR" > message.txt
+      env:
+        PR_TITLE: ${{fromJson(steps.get_pr.outputs.data)[0].title}}
+        PR_URL: ${{fromJson(steps.get_pr.outputs.data)[0].html_url}}
+        PR_AUTHOR: ${{fromJson(steps.get_pr.outputs.data)[0].user.login}}
+        PR_NUMBER: ${{fromJson(steps.get_pr.outputs.data)[0].number}}
+    - name: Upload message
+      uses: actions/upload-artifact@v3
+      with:
+        name: ${{ env.commit_date }}
```
Comment (sunnyyang99) [inline .github/workflows/get_prs_in_release.yml:107]: By storing the message in a folder with the commit date, the slack message will be ordered by date ascending.

File: .github/workflows/get_prs_in_release.yml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        commit_date=$(echo ${{ matrix.commits.commit.author.date }} | tr -d "TZ:-")
+        echo "commit_date=$commit_date" >> $GITHUB_ENV
+        echo "• \`$PR_TITLE\` (<$PR_URL|$PR_NUMBER>) by $PR_AUTHOR" > message.txt
+      env:
+        PR_TITLE: ${{fromJson(steps.get_pr.outputs.data)[0].title}}
+        PR_URL: ${{fromJson(steps.get_pr.outputs.data)[0].html_url}}
+        PR_AUTHOR: ${{fromJson(steps.get_pr.outputs.data)[0].user.login}}
+        PR_NUMBER: ${{fromJson(steps.get_pr.outputs.data)[0].number}}
+    - name: Upload message
+      uses: actions/upload-artifact@v3
```
Comment (sunnyyang99) [inline .github/workflows/get_prs_in_release.yml:105]: There has been no progress to access all outputs of matrix jobs. The workaround is to store the result in artifacts, which can be accessed by a different job.
https://github.com/orgs/community/discussions/17245

File: .github/workflows/get_prs_in_release.yml [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+  get_prs:
+    needs: get_commits_in_release
+    runs-on: ubuntu-latest
+    if: ${{ needs.get_commits_in_release.outputs.commits != '[]' && needs.get_commits_in_release.outputs.commits != '' }}
+    strategy:
+      matrix:
+        commits: ${{fromJson(needs.get_commits_in_release.outputs.commits)}}
+    steps:
+    - name: Auth with GCP
```
Comment (sunnyyang99) [inline .github/workflows/get_prs_in_release.yml:75]: I wasn't sure if I should have a job that sets the tokens as outputs, so each job doesn't have to have the Auth and Fetch secrets step.

File: .github/workflows/deploy_test.yml
Comment (jsonreeder) [inline .github/workflows/deploy_test.yml:1]: It's nice to see that you have created this for testing. I would expect it to be removed before merging.

**Review by jsonreeder — COMMENTED** [review]
I think there will likely be overlap with this other issue of creating GitHub "releases" (https://app.asana.com/0/1204858156175152/1204885126262733/f). We might, for example, set up a CI job to write all of this info to a GitHub release and then just link that in Slack.

**Comment by jsonreeder** [comment]
You've used `release` as language here. It makes some sense, though I have a preference for using "release" to talk about a product change and "deploy" to talk about a code change. Those things are often but not always the same. Here I think "deploy" may be a better fit.

**Review by jsonreeder — APPROVED** [review]
> Next steps would be integrating this with the API, FE, and ETL project.

I'd recommend starting to integrate with just one of these. Let's see how it looks. I've shared some high level questions and hesitations in comments, but I'm approving because I think this is a step in a good direction. I think seeing how it looks in practice when called will help me envision this system.

**Comment by sunnyyang99** [comment]
Thanks Jason! I will make the following changes before merging and beginning integration with FE.
- Rename file to `get_prs_in_deploy`
- Adjust workflow to output the desired message and not post to slack
- Remove test file

**Comment by leebs778** [comment]
+1 to Jasons comments

**Review by leebs778 — APPROVED** [review]

**Pull request merged**